content/blog/how-tikv-accesses-data.md

```diff
@@ -0,0 +1,263 @@
+
+{{< figure
+    src="/img/blog/how-tikv-accesses-data/rawkv-write.png"
+    caption="Write with RawKV"
+    number="" >}}
+
+A regular write operation with the RawKV API, writing `a = 1` for example, involves the following steps:
+
+1. Client requests the Region for a from PD
+2. PD returns the Region related information, mainly the TiKV node where the Leader is located
+3. Client sends command to the TiKV node where Leader is located
+4. Leader accepts the request and executes Raft process
+5. Leader applies `a = 1` to KV RocksDB and returns write success to Client
+
+### Read with RawKV
+
+{{< figure
+    src="/img/blog/how-tikv-accesses-data/rawkv-read.png"
+    caption="Read with RawKV"
+    number="" >}}
+
+Read is similar with write. The only difference is that Leader can directly apply a read without going through the Raft process.
+
+## TxnKV
+
+TxnKV corresponds to the Percolator mentioned above. It is a higher-level key-value API that provides ACID semantics. TxnKV is used when your application requires distributed transactions and/or MVCC.
+
+
+### Write with TxnKV
+
+{{< figure
+    src="/img/blog/how-tikv-accesses-data/txnkv-write.png"
+    caption="Write with TxnKV"
+    number="" >}}
+
+For TxnKV, the situation is much more complicated. Most of the processes have been explained in the [Percolator](#percolator) section. For TiKV to quickly seek the latest commit, the following steps are performed:
+
+1. Use big-endian to convert the timestamp to a 8-byte value, and then invert the bytes by bit.
+2. Combine the inverted bytes with the original key for storage in RocksDB
+
+This way the latest commits can be stored in the front, and the seek can be directly located. Of course, the corresponding TS is also encoded and processed in the same way.
+
+For example, suppose a key now has two commits—commitTS is 10 and 12, and startTS is 9 and 11, then the order of keys in RocksDB is:
+
+```
+Write CF:
+
+a_12-> 11
+a_10-> 9
+
+Data CF:
+
+a_11-> data_11
+a_9-> data_9
+```
+
+In addition, it should be noted that for small values, TiKV directly stores the value in the Write CF, so that only Read CF is required when reading. For write operations, the process is as follows:
+
+```
+PreWrite:
+
+Lock CF: W a-> Lock + Data
+
+Commit:
+Lock CF: R a-> Lock + 10 + Data
+Lock CF: D a
+
+Write CF: W a_11-> 10 + Data
+```
+
+For TiKV, the Lock CF will be read during the Commit phase to determine transaction conflicts, so we can get data from the Lock CF and write it to the Write CF.
+
+### Read with TxnKV
+
+{{< figure
+    src="/img/blog/how-tikv-accesses-data/txnkv-read.png"
+    caption="Read with TxnKV"
+    number="" >}}
+
+Refer to the [Percolator](#percolator) section for detailed write process with the TxnKV.
+
+### SQL Key Mapping between TiDB and TiKV
+
+TiKV is a distributed KV store. On top of it we built TiDB, a distributed relational database. You may wonder how a relational table is mapped to key-value. Take the following table:
```

Comment:
```suggestion
TiKV is a distributed KV store. On top of it companies are building databases that speak different dialects. How can they use TiKV to do this? Let's take a look at TiDB, a distributed relational database from PingCAP that speaks the MySQL protocol.

You may wonder how a relational table is mapped to key-value. Take the following table:
```

Reply:
A very nice edit. I actually was not sure about how I word this so that it does not sound like an original article for TiDB.